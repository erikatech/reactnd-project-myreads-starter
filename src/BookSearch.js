import React, {Component} from 'react';
import BookShelf from './BookShelf';
import {Link} from 'react-router-dom';
import * as BooksAPI from './BooksAPI'
class BookSearch extends Component {

	constructor(props) {
		super(props);
		this.state = {
			books: [],
			term: ''
		}
	}

	search = (term) => {
		this.setState({term});
		BooksAPI.search(term, 20)
		  .then((books) => this.setState({books}));
	};

	render() {
		return (
		  <div className="search-books">
			  <div className="search-books-bar">
				  <Link
					to="/"
					className="close-search">
					  Close
				  </Link>
				  <div className="search-books-input-wrapper">
					  <input type="text"
					         value={this.state.term}
					         onChange={(e) => this.search(e.target.value)}
					         placeholder="Search by title or author"/>
				  </div>
			  </div>
			  <div className="search-books-results">
				  <ol className="books-grid">
					  <BookShelf
					    shelf='All'
					    books={this.state.books}
					    moveBook={this.props.moveBook}
					  />
				  </ol>
			  </div>
		  </div>
		)
	}

};
export default BookSearch;