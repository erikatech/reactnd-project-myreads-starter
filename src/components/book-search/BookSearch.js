import React, {Component} from 'react';
import BookShelf from '../book-shelf/BookShelf';
import * as BooksAPI from '../../BooksAPI';
import SearchBar from '../seach-bar/SearchBar';

/**
 * Component that represents the search page
 * It receive as props the currentBooks that are defined in App.js
 *  so we can know which searched books are already added to some shelf
 * It also receive as props the moveBook method, which will be passed to BookShelf component
 */
class BookSearch extends Component {

	// initializes the book list
	state = {
		books: []
	};

	/**
	 * Search a book according to the user input
	 * @param term the user input value
	 */
	bookSearch = (term) => {
		BooksAPI.search(term, 20).then((result) => {
			// if the result return error, it means it didn't find any book
			if(result.error){
				// so the book state will be set to an empty array
				this.setState({books: []});
			} else {
				// this API method doesn't return a book with a shelf,
				// so wee need to check if some book inside result
				// is already added to a shelf.
				// when we find it, we set the shelf to the book inside result.
				this.props.currentBooks.forEach(currentBook => {
					result
					  .filter(book => book.id === currentBook.id)
					  .map(_ => _.shelf = currentBook.shelf);
				});

				// updates the current state
				this.setState({books: result});
			}
		})
	};





	render() {
		const shelf = {
			books: this.state.books,
			shelf: 'All'
		};
		return (
		  <div className="search-books">
			  <SearchBar onSearchTermChange={term => this.bookSearch(term)} />
			  <div className="search-books-results">
				  <ol className="books-grid">
					  <BookShelf
						shelf={shelf}
						moveBook={this.props.moveBook}
					  />
				  </ol>
			  </div>
		  </div>
		)
	}
}
export default BookSearch;