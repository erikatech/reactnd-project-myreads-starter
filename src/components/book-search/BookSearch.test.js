import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import BookSearch from './BookSearch';

Enzyme.configure({ adapter: new Adapter() });

describe('BookSearch testing', () => {

	it('should render without crashing', () => {
		expect(shallow(<BookSearch/>)).toMatchSnapshot();
	})
});





