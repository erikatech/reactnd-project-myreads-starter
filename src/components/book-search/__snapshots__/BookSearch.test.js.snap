// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookSearch testing should render without crashing 1`] = `
<div
  className="search-books"
>
  <SearchBar
    onSearchTermChange={[Function]}
  />
  <div
    className="search-books-results"
  >
    <ol
      className="books-grid"
    >
      <BookShelf
        shelf={
          Object {
            "books": Array [],
            "shelf": "All",
          }
        }
      />
    </ol>
  </div>
</div>
`;
